crates/sim2h/src/lib.rs

```diff
@@ -903,369 +1145,73 @@ impl Sim2h {
-                                    entry_aspect: aspect,
-                                },
-                            ));
-                        }
+            let r = state
+                .get_gossip_aspects_needed_for_agent(&space_hash, &agent_id)
+                .unwrap();
 
-                        let store_message = WireMessage::MultiSend(multi_messages);
+            trace!("sim2h gossip entry count: {}", r.len());
 
-                        let mut state = sim2h_handle.lock_state().await;
-                        let maybe_url = state.lookup_joined(&space_address, &to_agent_id);
-                        if maybe_url.is_none() {
-                            error!("Got FetchEntryResult with request id that is not a known agent id. I guess we lost that agent before we could deliver missing aspects.");
-                            return;
-                        }
-                        let url = maybe_url.unwrap();
-                        for (entry_address, aspect_address) in to_remove.drain(..) {
-                            state.remove_missing_aspect(
-                                &space_address,
-                                &to_agent_id,
-                                &entry_address,
-                                &aspect_address,
-                            );
-                        }
-                        state.send(to_agent_id, url, &store_message);
-                    });
+            for (entry_hash, aspects) in r.iter() {
+                if aspects.is_empty() {
+                    continue;
                 }
 
-                Ok(())
-            }
-            WireMessage::ClientToLib3h(ClientToLib3h::QueryEntry(query_data)) => {
-                if let DhtAlgorithm::NaiveSharding {redundant_count} = self.dht_algorithm {
-                    let sim2h_handle = self.sim2h_handle.clone();
-                    tokio::task::spawn(async move {
-                        let disconnects = sim2h_handle
-                            .lock_state().await
-                            .build_query(
-                                space_address,
-                                query_data,
-                                redundant_count
-                            );
-                        sim2h_handle.disconnect(disconnects);
-                    });
-                    Ok(())
-                } else {
-                    Err("Got ClientToLib3h::QueryEntry in full-sync mode".into())
-                }
-            }
-            WireMessage::Lib3hToClientResponse(Lib3hToClientResponse::HandleQueryEntryResult(query_result)) => {
-                if (query_result.responder_agent_id != agent_id) || (query_result.space_address != space_address)
-                {
-                    return Err(SPACE_MISMATCH_ERR_STR.into());
+                let query_agents = state.get_agents_for_query(&space_hash, &entry_hash, None);
+
+                if query_agents.is_empty() {
+                    continue;
```

Comment:
Shouldn't we be logging this as a warning at least?  I think this means that we know that an agent needs a particular aspect, but there's nobody to ask for it about.  That's a problem that should be logged?

Reply:
:+1: added `warn!` we may find this is too chatty when running real apps and people are going offline, etc - but better too verbose to start : )